### PR TITLE
Pending reuse pool

### PR DESCRIPTION
**Pull request by ocrickard**
This is the counter-proposal to https://github.com/facebook/componentkit/pull/692.

The core idea here is that we introduce a pending reuse pool, and move the dispatch_async hack into the reuse pool. This will allow us to prefer reusing "fully" relinquished views, but will allow fallbacks in the case that no fully relinquished view is present.

This still needs a little bit of work, but here it is.

### CODE REVIEW

File: ComponentKit/StatefulViews/CKStatefulViewReusePool.mm [hunk 1]
```diff
@@ -10,40 +10,76 @@
...
+#import "CKAssert.h"
 
 #import <unordered_map>
 
+struct FBStatefulReusePoolItemEntry {
+  UIView *view;
+  CKStatefulViewReusePoolPendingMayRelinquishBlock block;
+};
+
 class FBStatefulReusePoolItem {
```
Comment (ocrickard) [inline ComponentKit/StatefulViews/CKStatefulViewReusePool.mm]: I decided to just reuse this custom multimap for my new pending pool and just add on the block API. For the normal pool, the block will be nil.

File: ComponentKit/StatefulViews/CKStatefulViewReusePool.h [hunk 2]
```diff
@@ -20,6 +22,7 @@
 
 - (void)enqueueStatefulView:(UIView *)view
          forControllerClass:(Class)controllerClass
-                    context:(id)context;
+                    context:(id)context
+            enqueueComplete:(CKStatefulViewReusePoolPendingMayRelinquishBlock)enqueueComplete;
```
Comment (adamjernst) [inline ComponentKit/StatefulViews/CKStatefulViewReusePool.h]: neat.

**Comment by adamjernst** [comment]
I like this quite a bit.

File: ComponentKit/StatefulViews/CKStatefulViewReusePool.h (hunk 2 above)
Comment (eczarny) [inline ComponentKit/StatefulViews/CKStatefulViewReusePool.h]: I'd name the parameter to better match how the block is being used. "Enqueue complete" sounds more informative than instructive. Maybe something that better captures the spirit of how the block is being used?

File: ComponentKit/StatefulViews/CKStatefulViewReusePool.mm [hunk 3]
```diff
@@ -10,40 +10,76 @@
...
+                           });
+    if (preferIt != _entries.end()) {
+      FBStatefulReusePoolItemEntry entry = *preferIt;
+      _entries.erase(preferIt);
+      if (entry.block == NULL || entry.block()) {
+        return entry.view;
+      }
+    }
+
+    if (_entries.empty()) {
```
Comment (eczarny) [inline ComponentKit/StatefulViews/CKStatefulViewReusePool.mm]: I'd include a comment explaining why we're checking if `_entries` is empty again. Even though there's a comment explaining what's going on just a bit further down.

At first I thought this was a mistake.

File: ComponentKit/StatefulViews/CKStatefulViewReusePool.mm [hunk 4]
```diff
@@ -10,40 +10,76 @@
...
   };
 
-  void addView(UIView *view)
+  void addEntry(const FBStatefulReusePoolItemEntry &entry)
   {
-    [views addObject:view];
+    _entries.push_back(entry);
   };
 
+  void absorbPendingPool(const FBStatefulReusePoolItem &otherPool, NSInteger maxEntries) {
```
Comment (eczarny) [inline ComponentKit/StatefulViews/CKStatefulViewReusePool.mm]: Nit: `{` on its own line.

File: ComponentKit/StatefulViews/CKStatefulViewReusePool.mm (hunk 3 above)
Comment (ocrickard) [inline ComponentKit/StatefulViews/CKStatefulViewReusePool.mm]: Turns out I didn't need this! If entries is empty, then entries.begin == entries.end, and we never enter the loop. I'm removing the check.

File: ComponentKit/StatefulViews/CKStatefulViewReusePool.mm [hunk 5]
```diff
@@ -10,40 +10,74 @@
...
   {
-    [views addObject:view];
+    _entries.push_back(entry);
   };
 
+  void absorbPendingPool(const FBStatefulReusePoolItem &otherPool, NSInteger maxEntries)
+  {
+    for (const FBStatefulReusePoolItemEntry &entry : otherPool._entries) {
+      if (entry.block == NULL || entry.block()) {
+        // The block was either not present, or evaluated to true, indicating it can join the normal reuse pool
```
Comment (eczarny) [inline ComponentKit/StatefulViews/CKStatefulViewReusePool.mm]: I'd rephrase this comment to mention why the entry can join the normal reuse pool. The logic is simple enough to follow so I don't see much of a need to reiterate it here. Why that condition is important is a bit more interesting.

File: ComponentKit/StatefulViews/CKStatefulViewReusePool.mm [hunk 6]
```diff
@@ -56,6 +90,8 @@ void addView(UIView *view)
 @implementation CKStatefulViewReusePool
 {
   std::unordered_map<std::pair<__unsafe_unretained Class, id>, FBStatefulReusePoolItem, PoolKeyHasher> _pool;
+  std::unordered_map<std::pair<__unsafe_unretained Class, id>, FBStatefulReusePoolItem, PoolKeyHasher> _pendingPool;
```
Comment (eczarny) [inline ComponentKit/StatefulViews/CKStatefulViewReusePool.mm]: Maybe name this to better indicate what's pending. `_pendingReusePool` is probably better. I may even rename `_pool` to `_reusePool` as well. Up to you.

File: ComponentKit/StatefulViews/CKStatefulViewReusePool.h [hunk 7]
```diff
@@ -10,16 +10,24 @@
...
+
 - (UIView *)dequeueStatefulViewForControllerClass:(Class)controllerClass
                                preferredSuperview:(UIView *)preferredSuperview
                                           context:(id)context;
 
 - (void)enqueueStatefulView:(UIView *)view
          forControllerClass:(Class)controllerClass
-                    context:(id)context;
+                    context:(id)context
+         mayRelinquishBlock:(CKStatefulViewReusePoolPendingMayRelinquishBlock)mayRelinquishBlock;
```
Comment (ocrickard) [inline ComponentKit/StatefulViews/CKStatefulViewReusePool.h]: Well, the view can say NO, and it won't be relinquished, so that's not representative. I think this name is fine. It's just an internal piece of infra, and is descriptive enough.

File: ComponentKit/StatefulViews/CKStatefulViewReusePool.mm (hunk 6 above)
Comment (ocrickard) [inline ComponentKit/StatefulViews/CKStatefulViewReusePool.mm]: Meh, this whole thing is a reuse pool. I don't think that improves readability at all. Up to you if you want to change it later though.

File: ComponentKit/StatefulViews/CKStatefulViewReusePool.mm (hunk 5 above)
Comment (ocrickard) [inline ComponentKit/StatefulViews/CKStatefulViewReusePool.mm]: Good point.

**Comment by ocrickard** [comment]
@facebook-github-bot import

**Comment by ocrickard** [comment]
@facebook-github-bot import

File: ComponentKit/StatefulViews/CKStatefulViewComponentController.mm [hunk 8]
```diff
@@ -131,18 +131,27 @@ - (void)_presentStatefulView
...
+         return YES;
+       }
+       if (!strongSelf->_mounted && [strongSelf canRelinquishStatefulView]) {
+         [strongSelf willRelinquishStatefulView:_statefulView];
+         strongSelf->_statefulView = nil;
+         strongSelf->_statefulViewContext = nil;
+         return YES;
+       }
+       return NO;
+     }];
```
Comment (eczarny) [inline ComponentKit/StatefulViews/CKStatefulViewComponentController.mm]: Another option is to just have this block call a private instance method so you can avoid the `weakSelf`/`strongSelf` dance.

File: ComponentKit/StatefulViews/CKStatefulViewReusePool.h (hunk 7 above)
Comment (eczarny) [inline ComponentKit/StatefulViews/CKStatefulViewReusePool.h]: Ah, good point. I was reading this block as being responsible for actually doing the relinquishing.

File: ComponentKit/StatefulViews/CKStatefulViewReusePool.mm [hunk 9]
```diff
@@ -10,40 +10,78 @@
...
 
+  void absorbPendingPool(const FBStatefulReusePoolItem &otherPool, NSInteger maxEntries)
+  {
+    for (const FBStatefulReusePoolItemEntry &entry : otherPool._entries) {
+      // In the future, we should consider not evaluating the block here immediately, and letting it move into the
+      // normal reuse pool. That way we can let stateful components hold onto their own views without any
+      // reconfiguration for a longer period of time.
+      if (entry.block == NULL || entry.block()) {
+        // The stateful view component can decide not to allow reuse of its view if the component has re-mounted before
+        // the block is evaluated.
```
Comment (eczarny) [inline ComponentKit/StatefulViews/CKStatefulViewReusePool.mm]: Awesome! Much clearer.

**Comment by clementgenzmer** [comment]
Does this fix the bug where a statefulView is to be recycled right now (in scroll view full of statefulView)?

**Comment by ocrickard** [comment]
@clementgenzmer yes it does, at the possible cost of more flickers on insertions above the viewport. We'll QE the change. There are several other things we can do if this isn't sufficient. In scrolling video home locally we're finally able to reuse views at a really high rate with this in place.

**Comment by ocrickard** [comment]
Ping @eczarny this has been updated.

**Comment by eczarny** [comment]
Let's try it out.

**Comment by ocrickard** [comment]
@facebook-github-bot import

**Comment by ocrickard** [comment]
Landing this internally, merging here.
